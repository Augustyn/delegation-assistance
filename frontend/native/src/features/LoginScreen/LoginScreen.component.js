import React from "react";
import { View, Button, StyleSheet } from "react-native";

const LoginScreen = props => {
  const { navigate } = props;
  return (
    <View style={styles.container}>
      <Button title="Sign in to app" onPress={() => navigate.navigate("Main")} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
    backgroundColor: "#f2f2f2"
  }
});

export default LoginScreen;
