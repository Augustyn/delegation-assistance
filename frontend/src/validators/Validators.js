const validateCheckInOutDate = (checkOutDate, allValues) => {
  if (checkOutDate <= allValues.checkInDate) {
    return "Check out date must be later than check in date!";
  }
  return null;
};

const validateDepartureArrivalDate = (arrivalDate, allValues) => {
  if (arrivalDate <= allValues.departureDate) {
    return "Arrival date must be later than departure date!";
  }
  return null;
};

const validateNumber = value => (value && isNaN(Number(value)) ? "Must be a number" : undefined);

const validateRequired = value => (value ? undefined : "This field is required.");

const validateStartEndDate = (endDate, allValues) => {
  if (endDate <= allValues.startDate) {
    return "End date must be later than start date!";
  }
  return null;
};

const validateCurrency = (value, { diet: perDiem }) => {
  if (value === undefined && perDiem !== undefined) {
    return "Diem must have currency.";
  }
  return null;
};

const validateDiem = (value, { diet: currency }) => {
  if (value === undefined && currency !== undefined) {
    return "Diem must be set.";
  }
  return null;
};

export {
  validateCheckInOutDate,
  validateDepartureArrivalDate,
  validateRequired,
  validateStartEndDate,
  validateNumber,
  validateCurrency,
  validateDiem
};
