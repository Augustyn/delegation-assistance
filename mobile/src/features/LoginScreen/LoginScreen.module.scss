@import "../../assets/styles/constants";

.container {
  flex: 1;
  align-items: center;
  justify-content: center;
  padding-bottom: 150px;
}

.title-text {
  padding-right: 60px;
  position: relative;
  font-size: 30px;
  font-weight: bold;
  margin: 10px;
  padding-left: 0;
}

.error-text {
  font-size: 13px;
  position: relative;
  text-align: center;
  color: red;
}

.frame-header {
  flex-direction: row;
}

.frame {
  margin: auto;
  align-items: center;
  border-width: 0.5px;
  border-color: $frame-border-color;
  width: 200px;
  height: 300px;
}

.login-container {
  margin: 10px;
}

.button-login {
  color: $primary-color;
}
