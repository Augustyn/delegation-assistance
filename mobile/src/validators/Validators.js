const validateRequired = value => (value ? undefined : "This field is required.");

const validateNumber = value => (value && isNaN(Number(value)) ? "Must be a number" : undefined);

const validateCurrency = value =>
  value && !/^[0-9]\d*(((,\d{3}){1})?(\.\d{0,2})?)$/i.test(value) ? "Must be money" : undefined;

const validateDiemExists = (value, { diet: currency }) => {
  if (value === undefined && currency !== undefined) {
    return "Diem must be set.";
  }
  return undefined;
};

const validateCurrencyExists = (value, { diet: perDiem }) => {
  if (value === undefined && perDiem !== undefined) {
    return "No currency.";
  }
  return undefined;
};

export { validateRequired, validateNumber, validateCurrency, validateDiemExists, validateCurrencyExists };
