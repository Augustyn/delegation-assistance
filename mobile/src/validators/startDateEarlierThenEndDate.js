const validate = (_, values) => {
  const errors = {};
  const startDate = values.dateNames[0];
  const endDate = values.dateNames[1];

  if (values.values[startDate] > values.values[endDate]) {
    errors[endDate] = "End date must be later than start date!";
  }

  return errors;
};

export default { validate };
