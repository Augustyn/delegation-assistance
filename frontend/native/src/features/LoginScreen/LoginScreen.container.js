import React, { Component } from "react";
import { connect } from "react-redux";

import { loginUser } from "../../actions/user.actions";
import { getLoggedStatus } from "../../selectors/user.selectors";
import LoginScreen from "./LoginScreen.component";

class LoginScreenContainer extends Component {
  static navigationOptions = {
    header: null
  };

  componentDidMount = () => {
    if (this.props.loggedStatus) {
      this.props.navigation.navigate("Main");
    }
  };

  render() {
    return <LoginScreen navigate={this.props.navigation} />;
  }
}

const mapStateToProps = state => {
  return {
    loggedStatus: getLoggedStatus(state)
  };
};

const mapDispatchToProps = {
  loginUser
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LoginScreenContainer);
