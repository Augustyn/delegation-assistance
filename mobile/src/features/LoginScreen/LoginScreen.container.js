import React, { Component } from "react";
import { connect } from "react-redux";

import { loginUser } from "../../actions/user.actions";
import { getLoggedStatus, getToken } from "../../selectors/user.selectors";
import LoginScreen from "./LoginScreen.component";

class LoginScreenContainer extends Component {
  static navigatonOptions = {
    header: null
  };

  constructor() {
    super();
    this.state = {
      errors: ""
    };
  }

  componentDidMount = () => {
    if (this.props.loggedStatus && this.props.myToken) {
      this.props.navigation.navigate("Main");
    }
    this.setState(() => {
      return {
        errors: ""
      };
    });
  };

  handleSubmit = values => {
    this.setState({ errors: "Loading..." });
    this.props
      .loginUser(values.login, values.password)
      .then(response => {
        if (response.status === 200) {
          this.props.navigation.navigate("Main");
        }
      })
      .catch(err => {
        this.setState({ errors: "Invalid Username or Password" });
      });
  };

  render() {
    return <LoginScreen onSubmit={this.handleSubmit} errors={this.state.errors} />;
  }
}

const mapStateToProps = state => {
  return {
    loggedStatus: getLoggedStatus(state),
    myToken: getToken(state)
  };
};

const mapDispatchToProps = {
  loginUser
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LoginScreenContainer);
