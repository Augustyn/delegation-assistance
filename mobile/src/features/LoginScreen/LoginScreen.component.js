import React from "react";
import { View, Text, Button } from "react-native";
import { Field, reduxForm } from "redux-form";

import { validateRequired } from "../../shared/Validators/Validators";
import FieldLoginRenderer from "../../components/FieldLoginRenderer/FieldLoginRenderer.component";
import styles from "./LoginScreen.module.scss";

const LoginScreen = props => {
  const { handleSubmit } = props;
  return (
    <View style={styles.container}>
      <View style={styles.frame}>
        <View style={styles["frame-header"]}>
          <Text style={styles["title-text"]}>Hello!</Text>
        </View>
        <View style={styles["login-container"]}>
          <Text style={styles["error-text"]}>{props.errors}</Text>
          <Field
            testID="field-login"
            name="login"
            placeholder="Username"
            component={FieldLoginRenderer}
            validate={[validateRequired]}
            isSecure={false}
          />
          <Field
            testID="field-password"
            name="password"
            placeholder="Password"
            component={FieldLoginRenderer}
            validate={[validateRequired]}
            isSecure={true}
          />
          <Button
            title="Sign in"
            onPress={handleSubmit}
            // TODO: export color scss variables from _constants
            // and import primary color here. (Example -> frontend/src/features/CreateDelegationPage/component)
            // The assignment below breaks tests because `styles` is mocked and
            // any atempts to access `styles` properties will
            // throw an exception. Any suggestions on how to resolve this problem
            // differently are welcome.
            // color={styles["button-login"].color}
            data-test="login-button"
          />
        </View>
      </View>
    </View>
  );
};

export default reduxForm({
  form: "loginform"
})(LoginScreen);
