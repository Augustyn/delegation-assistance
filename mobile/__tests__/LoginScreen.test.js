import React from "react";
import { View, Button, Text } from "react-native";
import { render, fireEvent, shallow } from "react-native-testing-library";
import { Field } from "redux-form";
import thunk from "redux-thunk";

import { validateRequired } from "../src/validators/Validators";
import LoginScreen from "../src/features/LoginScreen/LoginScreen.component";
import { LoginScreenContainer } from "../src/features/LoginScreen/LoginScreen.container";
import {
  renderWithTestingLibrary,
  renderWithRenderer
} from "../test/redux-mock.js";

describe("Login screen rendering", () => {
  let wrapper;

  beforeAll(() => {
    wrapper = renderWithTestingLibrary(<LoginScreen />);
  });

  it("renders the Login screen", async () => {
    const tree = renderWithRenderer(<LoginScreen handleSubmit={jest.fn()} />);
    expect(tree).toMatchSnapshot();
  });

  it("login should be required", () => {
    const { getByTestId } = wrapper;
    const field = getByTestId("field-login");
    expect(field).toBeDefined();
    expect(field.props.validate).toContain(validateRequired);
  });

  it("password should", () => {
    const { getByTestId } = wrapper;
    const field = getByTestId("field-password");
    expect(field).toBeDefined();
    expect(field.props.validate).toContain(validateRequired);
  });

  it("passwrod should be secure type", () => {
    const { getByTestId } = wrapper;
    const field = getByTestId("field-password");
    expect(field).toBeDefined();
    expect(field.props.isSecure).toBe(true);
  });

  it("should render hello message", () => {
    const { getByText } = wrapper;
    expect(getByText("Hello!")).toBeDefined();
  });
});

describe("Login screen interactions", () => {
  it("should call handleSubmit on submit", () => {
    const onSubmitMock = jest.fn();
    const { getByType } = renderWithTestingLibrary(
      <LoginScreen handleSubmit={onSubmitMock} />
    );
    const btn = getByType(Button);
    fireEvent(btn, "press");
    expect(onSubmitMock).toHaveBeenCalled();
  });
});
