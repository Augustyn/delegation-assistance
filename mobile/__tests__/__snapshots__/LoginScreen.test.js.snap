// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login screen rendering renders the Login screen 1`] = `
<View>
  <View>
    <View>
      <Text>
        Hello!
      </Text>
    </View>
    <View>
      <Text />
      <View>
        <View>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            name="login"
            onBlur={[Function]}
            onChangeText={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="Username"
            secureTextEntry={false}
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
      <View>
        <View>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            name="password"
            onBlur={[Function]}
            onChangeText={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="Password"
            secureTextEntry={true}
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
      <View
        accessibilityRole="button"
        accessibilityStates={Array []}
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Sign in
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
